jupyter_scheduler/executors.py

```diff
@@ -221,3 +222,66 @@ def execute(self):
             archive_filepath = self.staging_paths["tar.gz"]
             with fsspec.open(archive_filepath, "wb") as f:
                 f.write(fh.getvalue())
+
+
+class AllFilesArchivingExecutionManager(DefaultExecutionManager):
+    """Execution manager that archives all output files in and under the
+    output directory into a single archive file
+
+    Notes
+    -----
+    Should be used along with :class:`~jupyter_scheduler.scheduler.AllFilesArchivingScheduler`
+    as the `scheduler_class` during jupyter server start.
+    """
+
+    def execute(self):
+        job = self.model
+
+        with open(self.staging_paths["input"], encoding="utf-8") as f:
+            nb = nbformat.read(f, as_version=4)
+
+        if job.parameters:
+            nb = add_parameters(nb, job.parameters)
+
+        ep = ExecutePreprocessor(
+            kernel_name=nb.metadata.kernelspec["name"],
+            store_widget_state=True,
+        )
+
+        # Get the directory of the input file
+        working_dir = os.path.dirname(os.path.abspath(self.staging_paths["input"]))
+
+        try:
+            ep.preprocess(nb, {"metadata": {"path": working_dir}})
+        except CellExecutionError as e:
+            pass
+        finally:
+            # Create all desired output files, other than "input" and "tar.gz"
+            for output_format in job.output_formats:
+                if output_format == "input" or output_format == "tar.gz":
+                    pass
+                else:
+                    cls = nbconvert.get_exporter(output_format)
+                    output, resources = cls().from_notebook_node(nb)
+                    f = open(self.staging_paths[output_format], "wb")
+                    f.write(bytes(output, "utf-8"))
+                    f.close()
+
+            # Create an archive file of the staging directory for this run
+            # and everything under it
+            fh = io.BytesIO()
+            with tarfile.open(fileobj=fh, mode="w:gz") as tar:
+                for root, dirs, files in os.walk(working_dir):
+                    for file in files:
+                        with open(os.path.join(root, file)) as f:
+                            output = f.read()
+                        data = bytes(output, "utf-8")
+                        source_f = io.BytesIO(initial_bytes=data)
+                        # TODO: Include relative path?
+                        info = tarfile.TarInfo(file)
+                        info.size = len(data)
+                        tar.addfile(info, source_f)
```

Comment:
1. We should decide on whether we're OK with flattening all the side effect files and not preserving their directory structure. This will be a backwards-incompatible change, so we will want alignment on this.

2. The current implementation doesn't clean up side effect files generated under `<staging-dir>/<job-id>`. If we're OK with flattening all the side effect files, you could instead run the notebook under a temporary dir `<staging-dir>/<job-id>/.run` and delete all those files after building the tarball:

```py
working_dir = os.path.dirname(self.staging_paths["input"])
run_dir = os.path.join(working_dir, ".run")

        try:
            ep.preprocess(nb, {"metadata": {"path": run_dir }})
        except CellExecutionError as e:
            pass
        finally:
            # ... shutil.rmtree(run_dir)
```

I would also rename `working_dir` to `local_staging_dir` for clarity.